sphinx/util/logging.py

```diff
@@ -27,3 +291,127 @@ def is_suppressed_warning(type, subtype, suppress_warnings):
+        if self.app.warningiserror:
+            raise SphinxWarning(record.msg % record.args)
+        else:
+            return True
+
+
+class WarningLogRecordTranslator(logging.Filter):
+    """Converts a log record to one Sphinx expects
+
+    * Make a instance of SphinxWarningLogRecord
+    * docname to path if location given
+    """
+    def __init__(self, app):
+        # type: (Sphinx) -> None
+        self.app = app
+        super(WarningLogRecordTranslator, self).__init__()
+
+    def filter(self, record):  # type: ignore
+        # type: (SphinxWarningLogRecord) -> bool
+        if isinstance(record, logging.LogRecord):
+            record.__class__ = SphinxWarningLogRecord  # force subclassing to handle location
+
+        location = getattr(record, 'location', None)
+        if isinstance(location, tuple):
+            docname, lineno = location
+            if docname and lineno:
+                record.location = '%s:%s' % (self.app.env.doc2path(docname), lineno)
+            elif docname:
+                record.location = '%s' % self.app.env.doc2path(docname)
+            else:
+                record.location = None
+        elif location and ':' not in location:
+            record.location = '%s' % self.app.env.doc2path(location)
+
+        return True
+
+
+class ColorizeFormatter(logging.Formatter):
+    def format(self, record):
+        message = super(ColorizeFormatter, self).format(record)
+        color = getattr(record, 'color', None)
+        if color is None:
+            color = COLOR_MAP.get(record.levelno)
+
+        if color:
+            return colorize(color, message)
+        else:
+            return message
+
+
+class SafeEncodingWriter(object):
+    """Stream writer which ignores UnicodeEncodeError silently"""
+    def __init__(self, stream):
+        self.stream = stream
+        self.encoding = getattr(stream, 'encoding', 'ascii') or 'ascii'
+
+    def write(self, data):
+        try:
+            self.stream.write(data)
+        except UnicodeEncodeError:
+            # stream accept only str, not bytes.  So, we encode and replace
+            # non-encodable characters, then decode them.
+            self.stream.write(data.encode(self.encoding, 'replace').decode(self.encoding))
+
+    def flush(self):
+        if hasattr(self.stream, 'flush'):
+            self.stream.flush()
+
+
+def setup(app, status, warning):
+    # type: (Sphinx, IO, IO) -> None
+    """Setup root logger for Sphinx"""
+    logger = logging.getLogger()
+    logger.setLevel(logging.NOTSET)
+
+    # clear all handlers
+    for handler in logger.handlers[:]:
+        logger.removeHandler(handler)
+
+    info_handler = NewLineStreamHandler(SafeEncodingWriter(status))
+    info_handler.addFilter(InfoFilter())
+    info_handler.setLevel(VERBOSITY_MAP.get(app.verbosity))
+    info_handler.setFormatter(ColorizeFormatter())
+
+    warning_handler = WarningStreamHandler(SafeEncodingWriter(warning))
```

Comment:
logging.StremHandler which is base class of WarningSteamHandler is defined as 
```
class StreamHandler(Handler):
    def __init__(self, stream: Optional[IO[str]] = ...) -> None: ...
```
https://github.com/python/typeshed/blob/master/stdlib/2and3/logging/__init__.pyi#L355

And IO is defined with many abstract methods:
https://github.com/python/typing/blob/master/src/typing.py#L1994